src/shaders/fill_extrusion.vertex.glsl

```diff
@@ -155,25 +165,38 @@ void main() {
     y_ground += y_ground * 5.0 / max(3.0, top_height);
 #endif
     v_ao = vec3(mix(concave, -concave, start), y_ground, h - ele);
-    directional *= (1.0 + 0.05 * (1.0 - top_up_ny.y) * u_ao[0]); // compensate sides faux ao shading contribution
+    NdotL *= (1.0 + 0.05 * (1.0 - top_up_ny.y) * u_ao[0]); // compensate sides faux ao shading contribution
 
 #ifdef PROJECTION_GLOBE_VIEW
     top_height += u_height_lift;
 #endif
     gl_Position.z -= (0.0000006 * (min(top_height, 500.) + 2.0 * min(base, 500.0) + 60.0 * concave + 3.0 * start)) * gl_Position.w;
 #endif
 
-    // Assign final color based on surface + ambient light color, diffuse light directional, and light color
+    vec3 indirect = vec3(0.0);
+    vec3 direct = vec3(0.0);
+#ifdef LIGHTING_3D_MODE
+    v_color = apply_lighting(color, NdotL);
+#else
+    // Assign final color based on surface + ambient light color, diffuse light NdotL, and light color
     // with lower bounds adjusted to hue of light
     // so that shading is tinted with the complementary (opposite) color to the light color
-    v_color.rgb += clamp(color.rgb * directional * u_lightcolor, mix(vec3(0.0), vec3(0.3), 1.0 - u_lightcolor), vec3(1.0));
+    v_color.rgb += clamp(color.rgb * NdotL * u_lightcolor, mix(vec3(0.0), vec3(0.3), 1.0 - u_lightcolor), vec3(1.0));
+#endif
+    
     v_color *= u_opacity;
 
 #ifdef ZERO_ROOF_RADIUS
     v_roof_color = vec4(0.0, 0.0, 0.0, 1.0);
-    float roof_radiance = clamp(u_lightpos.z, 0.0, 1.0);
-    roof_radiance = mix((1.0 - u_lightintensity), max((1.0 - colorvalue + u_lightintensity), 1.0), roof_radiance);
-    v_roof_color.rgb += clamp(color.rgb * roof_radiance * u_lightcolor, mix(vec3(0.0), vec3(0.3), 1.0 - u_lightcolor), vec3(1.0));
+
+#ifdef LIGHTING_3D_MODE
+    v_roof_color = apply_lighting(color);
```

Comment:
Should you use NdotL with roof lighting as well? Now there might be a discontinuity between walls and roofs

Reply:
No this is fine. This code is used only when rounded edges are enabled but we'd like to keep the roof edges from getting rounded (it's under `ZERO_ROOF_RADIUS`).

Reply:
But perhaps I misunderstood your concern. Indeed for the roof top now the `calculate_NdotL` (with extended clamping) isn't used. This does create some sort of discontinuity with walls but it's consistent with the flat 2d layers.

Reply:
I think it's perfectly fine using (0, 0, 1) as the normal vector for roofs. Having wider shading range is relevant for walls as otherwise backfaces would look really dull but this is not the case with flat (z=0) planes. We can revisit this if the other option ends up being better.